File: Official Windows Platform Sample/Windows 8.1 Store app samples/99866-Windows 8.1 Store app samples/Responding to the appearance of the on-screen keyboard sample/C#/InputPaneHelper.cs
using System;
using System.Collections.Generic;
using Windows.UI.ViewManagement;
using Windows.UI.Xaml;
using Windows.Foundation;
using Windows.UI.Xaml.Media.Animation;

namespace KeyboardEventsSampleCS
{
    public delegate void InputPaneShowingHandler(object sender, InputPaneVisibilityEventArgs e);
    public delegate void InputPaneHidingHandler(InputPane input, InputPaneVisibilityEventArgs e);
    public class InputPaneHelper
    {
        private Dictionary<UIElement, InputPaneShowingHandler> handlerMap;
        private UIElement lastFocusedElement = null;
        private InputPaneHidingHandler hidingHandlerDelegate = null;

        public InputPaneHelper()
        {
            handlerMap = new Dictionary<UIElement, InputPaneShowingHandler>();
        }

        public void SubscribeToKeyboard(bool subscribe)
        {
            InputPane input = InputPane.GetForCurrentView();
            if (subscribe)
            {
                input.Showing += ShowingHandler;
                input.Hiding += HidingHandler;
            }
            else
            {
                input.Showing -= ShowingHandler;
                input.Hiding -= HidingHandler;
            }
        }

        public void AddShowingHandler(UIElement element, InputPaneShowingHandler handler)
        {
            if (handlerMap.ContainsKey(element))
            {
                throw new System.Exception("A handler is already registered!");
            }
            else
            {
                handlerMap.Add(element, handler);
                element.GotFocus += GotFocusHandler;
                element.LostFocus += LostFocusHandler;
            }
        }

        // It's important to tell which element got focus in order to decide if
        // the keyboard handler should be called
        private void GotFocusHandler(object sender, RoutedEventArgs e)
        {
            lastFocusedElement = (UIElement)sender;
        }

        private void LostFocusHandler(object sender, RoutedEventArgs e)
        {
            if (lastFocusedElement == (UIElement)sender)
            {
                lastFocusedElement = null;
            }
        }

        private void ShowingHandler(InputPane sender, InputPaneVisibilityEventArgs e)
        {
            if (lastFocusedElement != null && handlerMap.Count > 0)
            {
                handlerMap[lastFocusedElement](lastFocusedElement, e);
            }
            lastFocusedElement = null;
        }

        private void HidingHandler(InputPane sender, InputPaneVisibilityEventArgs e)
        {
            if (hidingHandlerDelegate != null)
            {
                hidingHandlerDelegate(sender, e);
            }
            lastFocusedElement = null;
        }

        public void SetHidingHandler(InputPaneHidingHandler handler)
        {
            this.hidingHandlerDelegate = handler;
        }

        public void RemoveShowingHandler(UIElement element)
        {
            handlerMap.Remove(element);
            element.GotFocus -= GotFocusHandler;
            element.LostFocus -= LostFocusHandler;
        }
    }
}

File: Official Windows Platform Sample/Windows 8 app samples/[VB]-Windows 8 app samples/VB/Windows 8 app samples/Responding to the appearance of the on-screen keyboard sample (Windows 8)/VB/InputPaneHelper.vb
Imports Windows.UI.ViewManagement
Imports Windows.UI.Xaml
Imports Windows.Foundation
Imports Windows.UI.Xaml.Media.Animation
Imports System.Collections.Generic

Public Delegate Sub InputPaneShowingHandler(sender As Object, e As InputPaneVisibilityEventArgs)
Public Delegate Sub InputPaneHidingHandler(input As InputPane, e As InputPaneVisibilityEventArgs)

Public Class InputPaneHelper
    Private handlerMap As Dictionary(Of UIElement, InputPaneShowingHandler)
    Private lastFocusedElement As UIElement = Nothing
    Private hidingHandlerDelegate As InputPaneHidingHandler = Nothing

    Public Sub New()
        handlerMap = New Dictionary(Of UIElement, InputPaneShowingHandler)
    End Sub

    Public Sub SubscribeToKeyboard(subscribe As Boolean)
        Dim input As InputPane = InputPane.GetForCurrentView

        If subscribe Then
            AddHandler input.Showing, AddressOf ShowingHandler
            AddHandler input.Hiding, AddressOf HidingHandler
        Else
            RemoveHandler input.Showing, AddressOf ShowingHandler
            RemoveHandler input.Hiding, AddressOf HidingHandler
        End If
    End Sub

    Public Sub AddShowingHandler(element As UIElement, handler As InputPaneShowingHandler)
        If handlerMap.ContainsKey(element) Then
            Throw New System.Exception("A handler is already registered!")
        Else
            handlerMap.Add(element, handler)
            AddHandler element.GotFocus, AddressOf GotFocusHandler
            AddHandler element.LostFocus, AddressOf LostFocusHandler
        End If
    End Sub

    ' It's important to tell which element got focus in order to decide if
    ' the keyboard handler should be called
    Private Sub GotFocusHandler(sender As Object, e As RoutedEventArgs)
        lastFocusedElement = DirectCast(sender, UIElement)
    End Sub

    Private Sub LostFocusHandler(sender As Object, e As RoutedEventArgs)
        If lastFocusedElement Is DirectCast(sender, UIElement) Then
            lastFocusedElement = Nothing
        End If
    End Sub

    Private Sub ShowingHandler(sender As InputPane, e As InputPaneVisibilityEventArgs)
        If lastFocusedElement IsNot Nothing AndAlso handlerMap.Count > 0 Then
            handlerMap(lastFocusedElement)(lastFocusedElement, e)
        End If
        lastFocusedElement = Nothing
    End Sub

    Private Sub HidingHandler(sender As InputPane, e As InputPaneVisibilityEventArgs)
        hidingHandlerDelegate(sender, e)
        lastFocusedElement = Nothing
    End Sub

    Public Sub SetHidingHandler(handler As InputPaneHidingHandler)
        Me.hidingHandlerDelegate = handler
    End Sub

    Public Sub RemoveShowingHandler(element As UIElement)
        handlerMap.Remove(element)
        RemoveHandler element.GotFocus, AddressOf GotFocusHandler
        RemoveHandler element.LostFocus, AddressOf LostFocusHandler
    End Sub
End Class
